src/plugins/asp/asp_thread.cpp

```diff
@@ -0,0 +1,72 @@
+/***************************************************************************
+ *  asp_thread.cpp - ASP environment providing Thread
+ *
+ *  Created: Thu Oct 20 15:49:31 2016
+ *  Copyright  2016 Björn Schäpers
+ *
+ ****************************************************************************/
+
+/*  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Library General Public License for more details.
+ *
+ *  Read the full text in the LICENSE.GPL file in the doc directory.
+ */
+
+#include "asp_thread.h"
+
+#include <clingo.hh>
+
+using namespace fawkes;
+
+/** @class ASPThread "clips_thread.h"
+ * ASP environment thread.
+ *
+ * @author Björn Schäpers
+ *
+ * @property ASPThread::asp_inifin_
+ * @brief The initializer/finalizer for the ASPAspect.
+ *
+ * @property ASPThread::clingo_mgr_inifin_
+ * @brief The initializer/finalizer for the ClingoManagerAspect.
+ *
+ * @property ASPThread::control_mgr_
+ * @brief The clingo control manager.
+ */
+
+/** Constructor. */
+ASPThread::ASPThread()
+: Thread("ASPThread", Thread::OPMODE_WAITFORWAKEUP),
+  AspectProviderAspect([this]() {
```

Comment:
The aspect provider aspect requires a list of providers. That list has to be available already at initialization time of the ctor and cannot be created in the ctor itself. That makes it tricky to create this list. The typical pattern is to have a member function which generates this list,. For an example, cf. the clips plugin:
https://github.com/fawkesrobotics/fawkes/blob/6bca98b65b08555bd56f9ff4338111a247e13eda/src/plugins/clips/clips_thread.cpp#L42
https://github.com/fawkesrobotics/fawkes/blob/6bca98b65b08555bd56f9ff4338111a247e13eda/src/plugins/clips/clips_thread.cpp#L96-L104

Here, Bjoern implemented this as a lambda function. Note the `()` after the lambda's closing `}`, which immediately executes the function. This avoids the need to specify a member to do this.

Nowadays, it could be done using an initializer list (which was not available when we added this pattern and Bjoern might not have been aware of it and at least used an anonymous lambda rather than a member).

Reply:
I have just pushed https://github.com/fawkesrobotics/fawkes/pull/23/commits/1d2271d6ba75b58720ba80cbde5871da7a2b9b12 which switches to the simpler initializer list.